### ISSUE TITLE
content: conflicting events which change fleet "Large Free Worlds"

### ISSUE THREAD

**Issue by EndrosG** [endless-sky/endless-sky#2549]
After creating my PR #2548 I checked if I could create a similar one for the fleets. It turned out that while it will be possible to *add variants*, it is hard or almost impossible to remove or modify an existing variant as they dont have unique names and coding a proper search pattern would be more than I am willing to do right now.

In my opinion, a *add variants to a fleet definition* feature would be very nice for any content creators and could also be used to really shorten the [`events.txt`](https://github.com/endless-sky/endless-sky/blob/master/data/events.txt) data file.
Because only the additional new content would have to be specified and not the old content.

I checked some content of the events.txt data file and found two "conflicting" events:
`event "southern carriers 3"` modifies "Large Free Worlds" in (Line 740)[https://github.com/endless-sky/endless-sky/blob/master/data/events.txt#L740] and 
`event "plasma turret available"` modifies the same "Large Free Worlds" in (Line 1606)[https://github.com/endless-sky/endless-sky/blob/master/data/events.txt#L1606].
While the first one extends the original definition of "Large Free Worlds" by four variants with combinations of Roost, Skein, Finch and Boxwing; the second one uses the later (in `event "dreadnought deployment"`) used definition.

If the plasma turret event happens before the southern carriers 3 event, the "Large Free Worlds" fleet will have a variant with Falcon (Plasma). If the plasma turret event comes after the southern carriers 3 event, the fleet will have the Falcon (Plasma), but no carriers anymore.
This is only resolved by the introduction of the Dreadnought, which provides a new definition.

Having a look at my pilot which I started in early 2016, I have an even more annoying state.
I had the plasma turret event in early 2016, then the dreadnought event also in early 2016 (i.e. without the carriers) and then, after updating somethen to v0.9.6 and playing again, I got the southern carriers 3 event patched in and that fleet definition is my current "Large Free Worlds" definition.
I.e. even after finishing all the main plot, the wanderers middle etc. and having all the regular events, currently my "Large Free Worlds" fleet definition does not feature the Falcon (Plasma) nor the Dreadnought, but only the Roost and Skein carriers.

Three things ToDo:
- [ ] introduce a "variant add" enhancement similar to #2548 
- [ ] modify the events which alter fleets to make use of that feature
- [ ] check the way how things are patched, because applying "southern carriers 3" after "dreadnought deployment" is definetively *wrong* by some means.

**Comment by Amazinite** [comment]
So fleets would essentially work in a similar fashion to outfitters and shipyards on planets, or just outfitters and shipyards in general? When adding a shipyard to a planet or an outfit to a ship, one must only do the following:
```
planet "Earth"
	shipyard "Deep Advanced"

shipyard "Deep Advanced"
	"Vanguard"
```
And by using the `clear` command, one can clear out an entire outfitter/shipyard or clear the shipyards from a planet.
```
planet "Earth"
	shipyard clear

shipyard "Deep Advanced"
	clear
	"Bactrian"
```

To have fleets work the same way would be great.

**Comment by Amazinite** [comment]
I'll just say that I remember there being an issue on this in the past, but I'm not sure where it's at (or if ti was just a discussion on Steam).

**Comment by tehhowch** [comment]
As written, fleets would be reset by the current specification format (e.g. backwards compatibility), while use of `"variant add" ____` instead of `variant ____` would insert ____ into the existing fleet list.

**Comment by Amazinite** [comment]
Found it #1520
Make sure you hit the points brought up here.

**Comment by EndrosG** [comment]
@Amazinite :
1. I said I would not code a "remove" variant thing, so
> possibly also adding an option to remove with variant -1 or similar?
from the original post is not part of this, as written before.
2. I do not see the exact compatibility issue with existing savegames. Unlike #1520, I define a new keyword "variant add" instead of overwriting the existing "variant" keyword.
  - Thus, loading an existing savegame will have exactly the same result as before.
  - Old and new keywords can be used at the same time.
    - If in the savegame an event with the new "variant add" keyword appears after another event changing the same fleet, it will just add that content.
    - If an oldstyle fully replacement appears after a new "variant add" modification, it will just clear the fleet definition as it used to do before. I.e. we can still have the new compact, shortened "Large Free Worlds" definition from "dreadnought deployment" with much less variants which would then overwrite the twice modified original LFW definition (modification 1: southern carriers 3; modification 2: plasma turret available).
3. Content creators will still be in charge to somehow take care what happens with their fleet definition when other events happen. This responsibility does not change.
4. This introduces a chance to reduce content conflicts, but it will not avoid them completely.

I think, except for the "remove" thing I covered all of #1520. If you think different, *please* say so.

**Comment by endless-sky** [comment]
Adding fleets to a variant without replacing the old ones is now possible as of 9d8f06b, but going through all the old data files and changing that could take some time.

**Comment by jafdy** [comment]
@endless-sky Is that just using `add variant` where `variant` normally is?

**Comment by tehhowch** [comment]
@jafdy , yes. the line would start `add variant ...` instead of `variant ...`
`variant ...` will now (as it did before) trigger a clearing of the existing fleet implementation

So if you are planning to refactor the game events, you would drop all repeated definitions and then prepend `add ` to the variant lines that introduce the new ship. For events that remove a ship from a fleet, you'll (currently) still need to fully define the entire fleet. If we get #2552 , then we can also create the `remove variant ...` syntax and further refactor events.

**Comment by endless-sky** [comment]
The "variant" in #2552 is actually something completely different - that's for variants of a ship model, not variants of a fleet definition.

There are very few cases where ships actually need to be removed from a fleet definition. In the existing game events, variants are sometimes removed just to keep the fleet definition from getting super long, but using "add" will eliminate the need to do that. (It was really just for the sake of avoiding huge blocks of text in the missions - there's no real performance penalty to having more variants in a fleet.)

**Comment by tehhowch** [comment]
In the base game, I agree (currently). For mods and custom content, I think it is important to offer this flexibility. It also prevents multiple mods from squashing each other's updated fleet definitions depending on load order.

**Comment by endless-sky** [comment]
Requiring every single variant in a fleet to have a user-specified unique identifier would be a non-starter - adding far more complexity than it's worth. The only realistic option would be to let you remove a variant by specifying its complete contents. For example, you could write:

```
remove variant
    "Star Queen"
    "Sparrow" 2
```

Then, if there was a variant with exactly 1 Star Queen and 2 Sparrows, it would be removed. (Otherwise, nothing would happen.)

**Comment by tehhowch** [comment]
I think that would be fantastic to be able to do. Modders will certainly know the additions they make to a fleet definition, and thus will be able to curate the fleet definitions the player experiences  as their storylines progress.

**Comment by tehhowch** [comment]
I think the only unresolved part of this issue is that the event order of the save file overrides the event order from the current game's data files, and I'm not sure that warrants alteration. I think this can be closed with milestone v0.9.7.

A possible fix would be to implement a `-` operator for system classes that are changed by GameEvents such that, when saving the game, the game writes the save file in such a manner that only the end state of the save game is recreated. I.E. loading the game would process a single "event", in which the accumulated total of all the past events is faithfully restored, rather than repeating each and every event as the event was originally written. That fix, however, seems like a totally new issue
